/v3/embeddings endpoint (#2724)
openvinotoolkit/model_server: demos/embeddings/README.md

        }
    ]
}
```

## Start-up
```bash
docker run -d --rm -p 8000:8000 -v $(pwd)/:/workspace:ro openvino/model_server:latest --port 9000 --rest_port 8000 --config_path /workspace/config.json --cpu_extension /ovms/lib/libopenvino_tokenizers.so
```
In case you want to use GPU device to run the embeddings model, add extra docker parameters `--device /dev/dri --group-add=$(stat -c "%g" /dev/dri/render* | head -n 1)` 
to `docker run` command, use the image with GPU support and make sure set the target_device in subconfig.json to GPU. 
Also make sure the export model quantization level and cache size fit to the GPU memory.
```

Wait for the model to load. You can check the status with a simple command:
```bash
curl -s http://localhost:8000/v1/config | jq -c .
{"Alibaba-NLP/gte-large-en-v1.5":{"model_version_status":[{"version":"1","state":"AVAILABLE","status":{"error_code":"OK","error_message":"OK"}}]},"embeddings_model":{"model_version_status":[{"version":"1","state":"AVAILABLE","status":{"error_code":"OK","error_message":"OK"}}]},"tokenizer":{"model_version_status":[{"version":"1","state":"AVAILABLE","status":{"error_code":"OK","error_message":"OK"}}]}}
```

## Client code

> A single servable exposes both `chat/completions` and `completions` endpoints with and without stream capabilities.

Reviewer: ```suggestion
A single servable exposes both `chat/completions` and `completions` endpoints, with and without stream capabilities.
```

> Chat endpoint is expected to be used for scenarios where conversation context should be pasted by the client and the model prompt is created by the server based on the jinja model template.

Reviewer: "The chat endpoint is designed for scenarios where the conversation context should be provided by the client, while the model prompt is generated by the server based on the Jinja model template."

> Chat endpoint is expected to be used for scenarios where conversation context should be pasted by the client and the model prompt is created by the server based on the jinja model template.

Reviewer: ```suggestion
"The chat endpoint is designed for scenarios where the conversation context should be provided by the client, while the model prompt is generated by the server based on the Jinja model template."
```

Author: To be changed in separate demo PR @dtrawins https://github.com/openvinotoolkit/model_server/pull/2728

> Completion endpoint should be used to pass the prompt directly by the client and for models without the jinja template.

Author: To be changed in separate demo PR @dtrawins https://github.com/openvinotoolkit/model_server/pull/2728

```bash
curl http://localhost:8000/v3/embeddings \
  -H "Content-Type: application/json" -d '{ "model": "Alibaba-NLP/gte-large-en-v1.5", "input": "hello world"}' | jq .
```
```json
{
  "object": "list",
  "data": [
    {
      "object": "embedding",
      "embedding": [
        -0.03440694510936737,
        -0.02553200162947178,
        -0.010130723007023335,
        -0.013917984440922737,
...
        0.02722850814461708,
        -0.017527244985103607,
        -0.0053995149210095406
      ],
      "index": 0
    }
  ]
}

```

> Altenratively there could be used openai python client like in the example below:

Reviewer: ```suggestion
Alternatively openai python client could be used, like in the example below:
```

Reviewer 2: whole paragraph should be dropped. Updated version is in https://github.com/openvinotoolkit/model_server/pull/2728/files

Author: To be changed in separate demo PR @dtrawins https://github.com/openvinotoolkit/model_server/pull/2728

```bash
pip3 install openai
```
```python
from openai import OpenAI

client = OpenAI(
  base_url="http://localhost:8000/v3",
  api_key="unused"
)
model = "Alibaba-NLP/gte-large-en-v1.5"
responses = client.embeddings.create(
    input=[
        "Hello my name is",
        "Model server can support embeddings endpoint"
    ],
    model=model,
)
for data in responses.data:
    print(data.embedding)
```

## Benchmarking feature extraction

> TBD

Reviewer: ?

# How to serve Embeddings models with via OpenAI API {#ovms_demos_embeddings}
This demo shows how to deploy embeddings models in the OpenVINO Model Server for text feature extractions.
Text generation use case is exposed via OpenAI API `embeddings` endpoint.

## Get the docker image

> Pull public image with CPU only support or including also GPU support.

Author: done
